server/src/index.ts

```diff
@@ -3,17 +3,81 @@
 // SPDX-License-Identifier: MIT
 
 import express, { Application, Request, Response } from 'express';
+import { validateLogin, performSignup, PlainPassword, UserRole } from './auth';
 import { TeacherCourses, getTeacherCourses } from './services/courses';
 import cors from 'cors';
 
 const app: Application = express();
 const parsedPort = Number(process.env.AALTO_GRADES_BACKEND_PORT);
 const port: number = isNaN(parsedPort) ? 3000 : parsedPort;
 
+interface LoginRequest {
+  username: string,
+  password: PlainPassword,
+}
+
+interface SignupRequest {
+  username: string,
+  password: PlainPassword,
+  email: string,
+  role: UserRole,
+}
+
+function validateUserRole(role: any): role is UserRole {
+  return typeof role === 'string' && (
+    role === 'Teacher' ||
+    role === 'Student' ||
+    role === 'Admin'
+  );
+}
+
+function validateLoginFormat(body: any): body is LoginRequest {
+  return body &&
+    body.username &&
+    body.password &&
+    typeof body.username === 'string' &&
+    typeof body.password === 'string';
+}
+
+function validateSignupFormat(body: any): body is SignupRequest {
```

Comment:
I wonder if these validate functions are really necessary, TypeScript may have some better way to achieve the same thing, but I'm not sure.

Reply:
As far as I can tell as well it doesn't seem like TypeScript itself offers any nice way to do this. However, in #116 I used [the yup library](https://github.com/jquense/yup) for validation, which in my opinion is nicer solution.